### PR TITLE
User add queue item

### PR DESCRIPTION
**Pull request by RoadBytes**
not perfect but learning

### CODE REVIEW

File: app/views/queue_items/index.html.haml [hunk 1]
```diff
@@ -18,7 +18,7 @@
             - @queue_items.each do |queue_item|
               %tr
                 %td
-                  %input.form-control(type="text" value="1")
+                  %input.form-control(type="text" value="#{queue_item.position}")
```
Comment (tomtomecek) [inline app/views/queue_items/index.html.haml]: I would recommend using `text_field_tag`. I don't prefer raw HTML, Rails helpers are more succinct.

File: app/views/videos/show.html.haml [hunk 2]
```diff
@@ -13,7 +13,7 @@
         %p= @video.description
         .actions
           %a.btn.btn-primary(href="") Watch Now
-          %a.btn.btn-default(href="") + My Queue
+          = link_to "+ My Queue", queue_items_path(video_id: @video.id, user_id: current_user.id), method: "POST", class: "btn btn-default"
```
Comment (tomtomecek) [inline app/views/videos/show.html.haml]: Passing `user_id` here is wrong. Your controller is aware of who is current_user. This only exposes it to the potential malicious user.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -19,4 +19,8 @@ def current_user
   def require_user
     block_client unless logged_in?
   end
+
+  def position_assignment
```
Comment (tomtomecek) [inline app/controllers/application_controller.rb]: When you look at how you named this method. `position` is always bound to queue item in this case. This kind of code - does not belong to ApplicationController - it is attached more to the model and its purpose is to determine position for newly created queue item. So you have 3 options, either 1. keep it in controller that uses it, 2. push it to User model, 3. push it to QueueItem model.

Controller should really handle things that are bound to request / response / params / flashes / sessions - this type of  method is more bound to the data layer, so if you want to refactor it - then pushing to the model would be your way. I would also recommend `my_queue_contains?(video)` to `User` model definitely. As basically the purpose of that method is to check whether that current_user has already queued the video? `current_user.has_already_queued?(video)`.

Now, what is the benefit of pushing it to the model layer? Making it an instance method - you will be able to use that method _everywhere_ where the instance will be present. Anywhere you can use `current_user` - which represents an instance of User class - you will be able to call the instance method. Thus - avoid refactoring code to `ApplicationController`.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery
 
-  helper_method :logged_in?, :current_user 
+  helper_method :logged_in?, :current_user, :position_assignment
```
Comment (tomtomecek) [inline app/controllers/application_controller.rb]: `:position_assignment`, well this should not be here. if you push the method to the model layer (preferably User) you will be able to use it _everywhere_ else the user instance is available.

File: app/controllers/queue_items_controller.rb [hunk 5]
```diff
@@ -5,9 +5,27 @@ def index
     @queue_items = current_user.queue_items
   end
 
+  def create
+    video = Video.find_by id: params[:video_id]
+
+    unless my_queue_contains?(video)
+      QueueItem.create(user_id:  current_user.id,
+                       video_id: params[:video_id],
+                       position: position_assignment)
```
Comment (tomtomecek) [inline app/controllers/queue_items_controller.rb]: I am not sure what is the benefit of using `user_id` and `video_id` over the methods defined by associations.

```ruby
QueueItem.create(user:  current_user,
                 video: video,
                 position: position_assignment)
```

You have `video` object available - which is by far better option than `params[:video_id]`.

File: spec/controllers/queue_items_controller_spec.rb [hunk 6]
```diff
@@ -44,10 +44,49 @@
...
+      expect(response).to redirect_to root_path
+    end
   end
 
-  it "redirects to root_path for unauthenticated users" do
-    delete :destroy, id: queue_item_one.id
-    expect(response).to redirect_to root_path
+  describe "POST #create" do
+    let(:authenticated_user) { create(:user) }
+    let(:queue_item_one)     { build(:queue_item, user: authenticated_user, video: create(:video)) }
```
Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: What is the reason you create an instance of a new `queue_item` in memory here? I would also use Fabrication over FactoryGirl.

File: app/controllers/queue_items_controller.rb (hunk 5 above)
Comment (RoadBytes) [inline app/controllers/queue_items_controller.rb]: Cool, yeah, good point.  I'm not sure why I saw it like that.  I guess I'll get the hang of the parts/methods/variables the more I notice my preferences.  Honestly, I didn't realize how hacky I was until you pointed it out here.  I'm just thinking getting a solution, and it's hard to see past my code.

File: spec/controllers/queue_items_controller_spec.rb (hunk 6 above)
Comment (RoadBytes) [inline spec/controllers/queue_items_controller_spec.rb]: Is there a specific reason for Fabrication?  They seem to be pretty much the same thing, bc I did a little reading and one didn't stand out too much.

Comment (tomtomecek) [inline spec/controllers/queue_items_controller_spec.rb]: This is from Kevin:


<p>
Kevin Wang (Instructor)
<br><br>
Mostly Fabrication is an object generation library that works with many ORMs such as ActiveRecord, Mongoid and Sequel and can work with pure Ruby whereas FactoryGirl is tied to Rails/ActiveRecord. Also I prefer the interfaces of Fabrication. That said, the two libraries have more in similar that they are different.
</p>

File: app/models/queue_item.rb [hunk 7]
```diff
@@ -8,4 +8,8 @@ def user_rating
     review = Review.where(user_id: user.id, video_id: video.id).first
     review.rating unless review.nil?
   end
+
+  def position_assignment
```
Comment (tomtomecek) [inline app/models/queue_item.rb]: But QueueItem class does not have access to `current_user`.

File: app/models/queue_item.rb [hunk 8]
```diff
@@ -8,4 +8,8 @@ def user_rating
...
+    current_user.queue_items.size + 1
```
Comment (tomtomecek) [inline app/models/queue_item.rb]: But QueueItem class does not have access to `current_user`.

**Pull request merged**